### PR TITLE
Doorkeeper `5.3` compatibility

### PR DESCRIPTION
**Pull request by gencer**
Uses doorkeeper `5.3` from `master`

### CODE REVIEW

**Comment by gencer** [comment]
@nbulaj before i continue to fix all errors (which is 6 at this time) I need to be sure of one thing.

Doorkeeper 5.3 uses polymorphic models. Can I use [Sequel Polymorphic](https://github.com/jackdempsey/sequel_polymorphic) extension here? Or should I continue with [Sequel's own path](https://sequel.jeremyevans.net/rdoc/files/doc/advanced_associations_rdoc.html#label-Polymorphic+Associations)

P.S.: This WIP also ignores all new doorkeeper specs resides in `controller/` and `oauth/` at the moment.

**Comment by nbulaj** [comment]
Hi @gencer . I thought I already used Sequel Polymorphic for `owner` config option... Check it please. BTW I'm for Sequel Polymorphic :)

**Comment by nbulaj** [comment]
Ah here it is https://github.com/nbulaj/doorkeeper-sequel/blob/master/doorkeeper-sequel.gemspec#L28

**Comment by gencer** [comment]
Why did I not notice this before ?!

Okay, I will go with it and update the PR as soon as possible.

**Comment by gencer** [comment]
@nbulaj checks are success. I just merged master branch of doorkeeper and applied all necessary changes.

File: Rakefile [hunk 1]
```diff
@@ -39,6 +39,8 @@ class ExtensionIntegrator
       FileUtils.rm("spec/lib/orm/active_record/stale_records_cleaner_spec.rb", force: true)
       # Remove generators specs because we are using our own
       FileUtils.rm(Dir.glob("spec/generators/*.rb"))
+      FileUtils.rm(Dir.glob("spec/controllers/*.rb"))
+      FileUtils.rm(Dir.glob("spec/lib/oauth/*.rb"))
```
Comment (nbulaj) [inline Rakefile]: :worried:

File: lib/doorkeeper-sequel.rb [hunk 2]
```diff
@@ -14,6 +14,7 @@
 require "doorkeeper-sequel/generators/migration_generator"
 require "doorkeeper-sequel/generators/previous_refresh_token_generator"
 require "doorkeeper-sequel/generators/confidential_applications_generator"
+require "doorkeeper-sequel/generators/previous_refresh_token_generator"
```
Comment (nbulaj) [inline lib/doorkeeper-sequel.rb]: I think it's for `PolymorphicResourceOwnerGenerator`, right? :)

File: lib/doorkeeper-sequel/generators/templates/enable_polymorphic_resource_owner_migration.rb [hunk 3]
```diff
@@ -0,0 +1,15 @@
+# frozen_string_literal: true
+
+Sequel.migration do
+  change do
+    alter_table(:oauth_access_grants) do
```
Comment (nbulaj) [inline lib/doorkeeper-sequel/generators/templates/enable_polymorphic_resource_owner_migration.rb:5]: Acess Tokens also

File: lib/doorkeeper-sequel/mixins/access_token_mixin.rb [hunk 4]
```diff
@@ -135,6 +151,10 @@ def secret_strategy
       def fallback_secret_strategy
         ::Doorkeeper.configuration.token_secret_fallback_strategy
       end
+
+	  def polymorphic_resource_owner?
+		columns.include?(:resource_owner_type)
```
Comment (nbulaj) [inline lib/doorkeeper-sequel/mixins/access_token_mixin.rb]: Code formatting here

File: lib/doorkeeper/orm/sequel/access_token.rb [hunk 5]
```diff
@@ -20,6 +20,10 @@ def created_at_desc
       def refresh_token_revoked_on_use?
         columns.include?(:previous_refresh_token)
       end
+	  
+	  def polymorphic_resource_owner?
+		columns.include?(:resource_owner_type)
```
Comment (nbulaj) [inline lib/doorkeeper/orm/sequel/access_token.rb]: Formatting here

File: Rakefile (hunk 1 above)
Comment (gencer) [reply Rakefile]: For all of your comments: I'm on it.

File: lib/doorkeeper-sequel/mixins/access_token_mixin.rb (hunk 4 above)
Comment (gencer) [reply lib/doorkeeper-sequel/mixins/access_token_mixin.rb]: used notepad here :)

File: Rakefile (hunk 1 above)
Comment (gencer) [reply Rakefile]: Re-added them and all green.

**Comment by gencer** [comment]
Hey @nbulaj, All of your comments and their fixes have been pushed. All seems green. Can you re-check?

--

**Update:** Found 2 issues and fixed them. (Formatting issues)

File: lib/doorkeeper-sequel/mixins/access_token_mixin.rb [hunk 6]
```diff
@@ -104,6 +110,7 @@ def scopes_match?(token_scopes, param_scopes, app_scopes)
       end
 
       def authorized_tokens_for(application_id, resource_owner_id)
+        resource_owner_id = resource_owner_id&.id if polymorphic_resource_owner? && !resource_owner_id.is_a?(Integer)
         where(application_id: application_id,
               resource_owner_id: resource_owner_id,
```
Comment (gencer) [inline lib/doorkeeper-sequel/mixins/access_token_mixin.rb]: @nbulaj, should we check type if its polymorphic? Tests are passed but what if an user provides an id and type but this id have 2 times with different types?

Example:
```
===========
resource_owner_id   resource_owner_type
     1                 User
     1                 Admin
===========
```

Comment (gencer) [reply lib/doorkeeper-sequel/mixins/access_token_mixin.rb]: Yes, we have. I checked on active_record orm library for it. I will add it now.

Comment (gencer) [reply lib/doorkeeper-sequel/mixins/access_token_mixin.rb]: Done.

**Comment by gencer** [comment]
@nbulaj, I was using some guard checks wrongly (was working anyway but not in correct way) i changed them according to Doorkeeper usage.

My changes are **done**. You can look it now and reply me if necessary or merge.

**Review by nbulaj — APPROVED** [review]
LGTM

:monorail: 

Thanks @gencer !

**Comment by gencer** [comment]
Thanks for the approval, @nbulaj! Can't wait to see it on RubyGems.

**Comment by nbulaj** [comment]
No no, thank you! I hope to release it tomorrow (March 10)

**Comment by nbulaj** [comment]
Released as 2.2.0

**Comment by nbulaj** [comment]
Hi @gencer . There are new changes in **master** branch for Doorkeeper that requires ORM refactoring too.

https://github.com/doorkeeper-gem/doorkeeper/pull/1387/files

I've already add the backward-compatible changes for mongodb extension, maybe you wanna do the same for this one? It's ok if not, I'll do it later.

https://github.com/doorkeeper-gem/doorkeeper-mongodb/commit/d9426a5d7ad3143a59082cb04f64f9b9323a920e#diff-2144eb91ee237cb5046683fbf684a929R172-R204

**Comment by gencer** [comment]
Hi @nbulaj. I'm on it. Will send PR soon.

**Pull request merged**